go/vt/vttablet/tabletserver/throttle/throttler.go

```diff
@@ -71,7 +72,7 @@ var (
 	sqlGrantThrottlerUser = []string{
 		`GRANT SELECT ON _vt.heartbeat TO %s`,
 	}
-	replicationLagQuery = `select unix_timestamp(now(6))-max(ts/1000000000) from _vt.heartbeat`
+	replicationLagQuery = `select unix_timestamp(now(6))-max(ts/1000000000) as replication_lag from _vt.heartbeat`
```

Comment:
Is this query semantically guaranteed to work, or could there possibly be implementation-specific race conditions in a sluggish scenario?

At least for MySQL (https://dev.mysql.com/doc/refman/5.6/en/date-and-time-functions.html#function_now):
> NOW() returns a constant time that indicates the time at which the statement began to execute.

What I don't know is whether a `SELECT`, outside of a transaction, with `autocommit=1`, will use a consistent snapshot for its read; and if so, whether that snapshot will be at the same time, or after, the value of `NOW()`.

Suppose there is no consistent snapshot, or there is but it comes from a timepoint later than `NOW()`. Then couldn't the query compute a timestamp, then wait a while due to sluggishness, then read `max(ts)` and see a value that is close enough to `NOW()`, and return 200 ok.

Whereas in the same scenario, if there was a consistent snapshot taken at `NOW()`, if there was lag, you would definitely see it, and not return 200 ok.

I'm just wondering if, to be guaranteed to detect sluggishness, whether this works, or if you need a different query such that the timestamp is guaranteed to be computed after `max(ts)`. e.g. something like
```sql
select max(ts/1000000000) as ts_max_heartbeat from _vt.heartbeat;
select unix_timestamp(SYSDATE(6)) as ts_now;
```

Here there's the risk that you could get false-negatives since the timestamp is computed after the heartbeat. But if things aren't sluggish and there isn't replica lag, hopefully the delay should be so minimal as to be inconsequential.

I'm far from an expert here, so I don't know if this line of reasoning is valid or not.

Reply:
Nice thinking! Experience with this type of query throughout the years (and specifically in my previous employer, as this is adapted from https://github.com/github/freno/), shows that this is not an issue.

One nice thing about this mechanism is that the 1sec threshold is not an absolute restriction. It is OK if we pass the threshold (indeed, the logic pretty much ensures we will -- e.g. we will be able to write to the server when we're at 0.99sec lag, which means our own write will quite possibly push lag beyond 1sec).

So, back to the scenario you depicted, let's say replica is lagging by 2sec and has slugishness of 1sec, which leads us to believe the lag is 0.99sec. This means we push forward. This means in the near future lag will be pushed to be above 2sec, at which time the 1sec slugishness is not sluggish enough to make us think the lag is < 1sec. So if the system remains at 1sec slugishness, we're at worst case lagging at 2sec.

Just to illustrate how crazy time evaluations are, you could use `SYSDATE()` and _still_ evaluate the wrong thing if the time is evaluated first, then slugishness begins, then evaluation of table row takes place.

Having said all that, I don't have a strong opinion either way; just my experience that this logic hasn't created any harm in my past experience.

Reply:
Thank you, I appreciate the explanation.

The case I'm still wondering about is the primary.

In your scenario you discussed a replica, where replication lag and sluggishness are both at play. Sluggishness can cause false positives, but if replication lag gets high enough, it won't matter.

On the primary, there is no replication lag to consider, only sluggishness. If the sluggishness of the `select from _vt.heartbeat` is equal to the sluggishness of `insert into _vt.heartbeat`, wouldn't they cancel out and always appear to never be any sluggishness? I guess the question is, is that possible, or in a high-sluggishness scenario, would sluggishness for an insert almost always be greater than the sluggishness of a select?

Anyway, I'm not concerned if you're not concerned.

Reply:
> The case I'm still wondering about is the primary.

I see what you mean, and your point is valid. I still don't know if slugishness on the primary won't affect the timestamp in the same way (that is, `CURDATE` is computed when MySQL is _ready to insert_, and then the INSERT stalls).

That is to say, _I don't know_, I'm speculating here; and because I haven't known it to be an issue (and I've worked on sluggish systems) I _assume_ this is not a big risk.

But point taken, let me look into it, I don't think there's any risk in converting the `insert` to use `curdate()`, in which case -- why not.

Reply:
> But point taken, let me look into it, I don't think there's any risk in converting the insert to use curdate(), in which case -- why not.

Do you mean `now()` or `systime()`?

Why would you switch the insert code (which, from what I can tell, currently gets the timestamp from go)? Wouldn't that potentially make the heartbeat less suitable for detecting sluggishness, since there's more of a chance that the timestamp might be computed close to when the insert commits, even in the face of sluggishness?

I'd guess that keeping what you have now - which has been used a lot in production already without known problems, and where in the worst case scenario it's equivalent to today's behavior which is to have no check-self at all - is probably better than introducing something we're less sure of that might introduce new false negatives.